### PR TITLE
Added support for new datatypes

### LINKED ISSUE #40: Add support for new types
- [x] Single option
- [x] Multi option
- [x] Date range
- [x] time range

The `date` and `time` data types are handled differently in CRM. The data must is always stored in the same format. But the `Time` field must only have a time picker and `Date` must only have date picker without time.

@AbhayVAshokan _a

### PR DESCRIPTION
**Pull request by AbhayVAshokan**
Fixes #40 
The following types have been added/modified:
1. `date`: Accepts date inputs without time using DatePicker.
2. `time`: Accepts time inputs without date using TimePicker.
3. `datetime`: Accepts date and time using DatePicker.
4. `date_range`: Accepts two dates using DatePicker.
5. `time_range`: Accepts two times using TimePicker.
6. `single_option`: Accepts a text using dropdown.
7. `multi_option`: Accepts an array of options using dropdown.

NOTE: The backend implementation of the same is a WIP.
Demo: https://vimeo.com/719024755/47b297eb86

### CODE REVIEW

**Comment by AbhayVAshokan** [comment]
@Amaljith-K _a please review.

File: src/components/FilterDropdown/SearchInput.jsx [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+      const [error1, error2] = error.toString().split(",");
+      const [value1, value2] = node.value
+        ? node.value.toString().split(",")
+        : ["", ""];
+
+      const handleRuleBetweenChange = (date, position) => {
+        const startDate = position === "start" ? date : value1;
+        const endDate = position === "end" ? date : value2;
+        handleChangeValue(`${startDate},${endDate}`);
+      };
```
Comment (Amaljith-K) [inline src/components/FilterDropdown/SearchInput.jsx:53]: https://vimeo.com/719093167/25ca1c423d

File: src/components/FilterDropdown/SearchInput.jsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        defaultValue={node.value}
+        placeholder="Select date"
+        showTime={node.type.match(DATETIME_TYPE_REGEX)}
+        onChange={({ $d }) => handleChangeValue(beginningOfDay($d))}
+        onKeyDown={handleKeyDown}
+      />
+    );
+  }
+
+  // type = "time" or type = "time_range"
```
Comment (Amaljith-K) [inline src/components/FilterDropdown/SearchInput.jsx:96]: This comment is redundant. It is obvious from the code.

File: src/components/FilterDropdown/SearchInput.jsx [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+      const [error1, error2] = error.toString().split(",");
+      const [value1, value2] = node.value
+        ? node.value.toString().split(",")
+        : ["", ""];
+
+      const handleRuleBetweenChange = (time, position) => {
+        const startTime = position === "start" ? time : value1;
+        const endTime = position === "end" ? time : value2;
+        handleChangeValue(`${startTime},${endTime}`);
+      };
```
Comment (Amaljith-K) [inline src/components/FilterDropdown/SearchInput.jsx:114]: Are we repeating code here?

File: src/components/FilterDropdown/utils.js [hunk 4]
```diff
@@ -1,16 +1,36 @@
...
+
+    if (isEmpty(value2)) error += ", This is a required field.";
+
+    if (value1 && value2 && new Date(value1) > new Date(value2)) {
+      error = "Start date must be before end date";
+    }
+
+    if (error) return error;
+
+    return "";
```
Comment (Amaljith-K) [inline src/components/FilterDropdown/utils.js:23]: Isn't this enough?
```suggestion
    return error;
```

File: src/components/utils.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      type.match(TIME_TYPE_REGEX) ||
+      type.match(TIME_RANGE_TYPE_REGEX)
+    ) {
+      nodeValue =
+        rule === "between"
+          ? value
+              .split(",")
+              .map(time => formatTime(time))
+              .join(" & ")
+          : formatTime(value);
```
Comment (Amaljith-K) [inline src/components/utils.js:44]: Can we move this to a function?
```js
nodeValue = applyFormatting(condition, formatTime);
```

File: src/constants.js [hunk 6]
```diff
@@ -27,15 +39,67 @@ export const LOGIC_LABELS = {
...
+  [ATTRIBUTE_TYPES.datetime]: [
+    "at",
+    "before",
+    "after",
+    "between",
+    "has_any_value",
+    "is_unknown",
+  ],
+  [ATTRIBUTE_TYPES.date_range]: ["between", "has_any_value", "is_unknown"],
+  [ATTRIBUTE_TYPES.time_range]: ["between", "has_any_value", "is_unknown"],
```
Comment (Amaljith-K) [inline src/constants.js:67]: I don't really understand how `between` rule works for time_range and date_range.
https://vimeo.com/719101765/57fb5c7cd3

File: src/constants.js [hunk 7]
```diff
@@ -27,15 +39,67 @@ export const LOGIC_LABELS = {
...
+
 export const TIME_TYPE_REGEX = /^(?:custom\.)?time$/g;
+
+export const DATE_RANGE_TYPE_REGEX = /^(?:custom\.)?date_range$/g;
+
+export const TIME_RANGE_TYPE_REGEX = /^(?:custom\.)?time_range$/g;
+
+export const SINGLE_OPTION_TYPE_REGEX = /^(?:custom\.)?single_option$/g;
+
+export const MULTI_OPTION_TYPE_REGEX = /^(?:custom\.)?multi_option$/g;
```
Comment (Amaljith-K) [inline src/constants.js:105]: https://vimeo.com/719102401/a24412f316

**Review by Amaljith-K — CHANGES_REQUESTED** [review]
@AbhayVAshokan _a

File: src/constants.js (hunk 6 above)
Comment (AbhayVAshokan) [reply src/constants.js:67]: @Amaljith-K yes, thats a great concern. the `between` filter might raise confusions. My idea was to only filter totally inclusive date or time ranges.
One other approach is to change the set of filters to something like `["starts_between", "ends_between", "has_any_value", "is_unknown"]`. WDYT?

**Comment by AbhayVAshokan** [comment]
@Amaljith-K _a

File: src/constants.js (hunk 6 above)
Comment (Amaljith-K) [reply src/constants.js:67]: Filter for range datatype should be a single date. 
The rule name can be something like: `date_in_range` (you can find something better).
The user will choose a date. We will find all rows whose date range includes the selected date.

**Comment by AbhayVAshokan** [comment]
@Amaljith-K I have updated the `between` rule to `contains` for both `date_range` and `time_range` types.

**Review by Amaljith-K — APPROVED** [review]
LGTM

**Pull request merged**